In [70]:
import warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split

### Acquire the train.csv

In [2]:
# Loade the csv files
train = pd.read_csv("train.csv")

# Inspect the shapes
train.shape

(101230332, 10)

In [3]:
# Take a quck peek of the dataframe
train.head()

row_id  timestamp  user_id  content_id  content_type_id  task_container_id  \
0       0          0      115        5692                0                  1   
1       1      56943      115        5716                0                  2   
2       2     118363      115         128                0                  0   
3       3     131167      115        7860                0                  3   
4       4     137965      115        7922                0                  4   

   user_answer  answered_correctly  prior_question_elapsed_time  \
0            3                   1                          NaN   
1            2                   1                      37000.0   
2            0                   1                      55000.0   
3            0                   1                      19000.0   
4            1                   1                      11000.0   

  prior_question_had_explanation  
0                            NaN  
1                          False  
2                          False  
3                          False  
4                          False

In [4]:
# Summarize the train dataset
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101230332 entries, 0 to 101230331
Data columns (total 10 columns):
 #   Column                          Dtype  
---  ------                          -----  
 0   row_id                          int64  
 1   timestamp                       int64  
 2   user_id                         int64  
 3   content_id                      int64  
 4   content_type_id                 int64  
 5   task_container_id               int64  
 6   user_answer                     int64  
 7   answered_correctly              int64  
 8   prior_question_elapsed_time     float64
 9   prior_question_had_explanation  object 
dtypes: float64(1), int64(8), object(1)
memory usage: 7.5+ GB


**Takeaways**
- The train dataset has more than 100 millions records. I only need to small portion to calculate the statistics of the past.<br>
    -**I will take the first 1000 user ids.**

In [8]:
# Extract the first 1000 user ids

user_ids1000 = train.user_id.value_counts().sort_index().iloc[0: 1000]
user_ids1000 = user_ids1000.index.to_list()

# Sanity check
print(len(user_ids1000))

# Print the first 5 user ids
user_ids1000[:5]

1000


[115, 124, 2746, 5382, 8623]

In [10]:
# Set the user_id to the index

train = train.set_index("user_id")
train.head()

row_id  timestamp  content_id  content_type_id  task_container_id  \
user_id                                                                      
115           0          0        5692                0                  1   
115           1      56943        5716                0                  2   
115           2     118363         128                0                  0   
115           3     131167        7860                0                  3   
115           4     137965        7922                0                  4   

         user_answer  answered_correctly  prior_question_elapsed_time  \
user_id                                                                 
115                3                   1                          NaN   
115                2                   1                      37000.0   
115                0                   1                      55000.0   
115                0                   1                      19000.0   
115                1                   1                      11000.0   

        prior_question_had_explanation  
user_id                                 
115                                NaN  
115                              False  
115                              False  
115                              False  
115                              False

In [11]:
# Subtracts the records for the first 1000 users
train1000 = train.loc[user_ids1000]

# Take a peek at the new dataset
train1000.head()

row_id  timestamp  content_id  content_type_id  task_container_id  \
user_id                                                                      
115           0          0        5692                0                  1   
115           1      56943        5716                0                  2   
115           2     118363         128                0                  0   
115           3     131167        7860                0                  3   
115           4     137965        7922                0                  4   

         user_answer  answered_correctly  prior_question_elapsed_time  \
user_id                                                                 
115                3                   1                          NaN   
115                2                   1                      37000.0   
115                0                   1                      55000.0   
115                0                   1                      19000.0   
115                1                   1                      11000.0   

        prior_question_had_explanation  
user_id                                 
115                                NaN  
115                              False  
115                              False  
115                              False  
115                              False

In [12]:
# Sanity check
train1000.tail() # should match the tail of user_ids1000

row_id  timestamp  content_id  content_type_id  task_container_id  \
user_id                                                                      
5577726  296610  625311764        5328                0                  7   
5577726  296611  625327721        4303                0                  8   
5577726  296612  625344560        4445                0                  9   
5577726  296613  625367161        5206                0                 10   
5577726  296614  625378759        3928                0                 11   

         user_answer  answered_correctly  prior_question_elapsed_time  \
user_id                                                                 
5577726            0                   0                      20000.0   
5577726            3                   0                      10000.0   
5577726            0                   1                       6000.0   
5577726            1                   0                       6000.0   
5577726            0                   0                      13000.0   

        prior_question_had_explanation  
user_id                                 
5577726                          False  
5577726                           True  
5577726                           True  
5577726                           True  
5577726                           True

In [14]:
# Print the tail of the user_ids1000
user_ids1000[999]

5577726

In [60]:
# Read the train1000.csv
train1000 = pd.read_csv("train1000.csv", index_col=0)

In [61]:
# Quick summarize the train1000
train1000.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 296615 entries, 115 to 5577726
Data columns (total 9 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   row_id                          296615 non-null  int64  
 1   timestamp                       296615 non-null  int64  
 2   content_id                      296615 non-null  int64  
 3   content_type_id                 296615 non-null  int64  
 4   task_container_id               296615 non-null  int64  
 5   user_answer                     296615 non-null  int64  
 6   answered_correctly              296615 non-null  int64  
 7   prior_question_elapsed_time     289945 non-null  float64
 8   prior_question_had_explanation  295617 non-null  object 
dtypes: float64(1), int64(7), object(1)
memory usage: 22.6+ MB


In [62]:
# Count how many missing values in the prior_question_elapsed_time

mask = train1000.prior_question_elapsed_time.isnull()
mask.sum()

6670

In [63]:
# Count how many missing values in the prior_question_had_explanation

mask = train1000.prior_question_had_explanation.isnull()
mask.sum()

998

**Takeaways**
- The memoery usage of the first 1000 users are only 23 MB
- Missing values are found in columns:
    - prior_question_elapsed_time: 6670 nulls
    - prior_question_had_explanantion: 998 nulls

### Handling the null values

In [64]:
# The values of the two prior columns will always be 0 when the user first interacted with the system
# Take a first user 115 for example

train1000.loc[115].head()

row_id  timestamp  content_id  content_type_id  task_container_id  \
user_id                                                                      
115           0          0        5692                0                  1   
115           1      56943        5716                0                  2   
115           2     118363         128                0                  0   
115           3     131167        7860                0                  3   
115           4     137965        7922                0                  4   

         user_answer  answered_correctly  prior_question_elapsed_time  \
user_id                                                                 
115                3                   1                          NaN   
115                2                   1                      37000.0   
115                0                   1                      55000.0   
115                0                   1                      19000.0   
115                1                   1                      11000.0   

        prior_question_had_explanation  
user_id                                 
115                                NaN  
115                              False  
115                              False  
115                              False  
115                              False

**Takeawasy**
- There are only 1000 users so the two prior columns should have at leat 1000 nulls.
- The prior_question_had_explanantion column has 998 nulls:
    - fill the nulls with 'False'
- I will further explore the missing values in the column prior_question_elapased_time.

### Fill the nulls in the prior_question_had_explanation with string 'False'

In [65]:
train1000.prior_question_had_explanation.fillna('False', inplace = True)
train1000.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 296615 entries, 115 to 5577726
Data columns (total 9 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   row_id                          296615 non-null  int64  
 1   timestamp                       296615 non-null  int64  
 2   content_id                      296615 non-null  int64  
 3   content_type_id                 296615 non-null  int64  
 4   task_container_id               296615 non-null  int64  
 5   user_answer                     296615 non-null  int64  
 6   answered_correctly              296615 non-null  int64  
 7   prior_question_elapsed_time     289945 non-null  float64
 8   prior_question_had_explanation  296615 non-null  object 
dtypes: float64(1), int64(7), object(1)
memory usage: 32.6+ MB


In [66]:
mask = train1000.prior_question_elapsed_time.isnull()
train1000[mask]

row_id   timestamp  content_id  content_type_id  task_container_id  \
user_id                                                                       
115           0           0        5692                0                  1   
124          46           0        7900                0                  0   
2746         76           0        5273                0                  0   
2746         89      653762        6808                1                 14   
5382         96           0        5000                0                  0   
...         ...         ...         ...              ...                ...   
5566848  296498   162115714       32168                1                 21   
5568049  296509           0        6028                0                  0   
5568049  296537  7746409318       24250                1                 20   
5571864  296586           0        4205                0                  0   
5577726  296603           0        6442                0                  0   

         user_answer  answered_correctly  prior_question_elapsed_time  \
user_id                                                                 
115                3                   1                          NaN   
124                0                   1                          NaN   
2746               1                   0                          NaN   
2746              -1                  -1                          NaN   
5382               0                   1                          NaN   
...              ...                 ...                          ...   
5566848           -1                  -1                          NaN   
5568049            3                   0                          NaN   
5568049           -1                  -1                          NaN   
5571864            1                   1                          NaN   
5577726            0                   0                          NaN   

        prior_question_had_explanation  
user_id                                 
115                              False  
124                              False  
2746                             False  
2746                             False  
5382                             False  
...                                ...  
5566848                          False  
5568049                          False  
5568049                          False  
5571864                          False  
5577726                          False  

[6670 rows x 9 columns]

**Takeaways**
- When the user studies the lecture, the prior_question_elapsed_time documented as NaN.
- I will fill the missing values in the columsn prior_question_elapsed_time with 0.

### Fill the nulls in the prior_question_elapsed_time with string 0

In [67]:
train1000.prior_question_elapsed_time.fillna(0, inplace = True)
train1000.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 296615 entries, 115 to 5577726
Data columns (total 9 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   row_id                          296615 non-null  int64  
 1   timestamp                       296615 non-null  int64  
 2   content_id                      296615 non-null  int64  
 3   content_type_id                 296615 non-null  int64  
 4   task_container_id               296615 non-null  int64  
 5   user_answer                     296615 non-null  int64  
 6   answered_correctly              296615 non-null  int64  
 7   prior_question_elapsed_time     296615 non-null  float64
 8   prior_question_had_explanation  296615 non-null  object 
dtypes: float64(1), int64(7), object(1)
memory usage: 32.6+ MB


**Takeawasy**
- All the null values have been addressed in the dataset. 

### Split the train1000 into train and test

In [69]:
# Double check the 1000 user_id list
print(len(user_ids1000))
user_ids1000[:5]

1000


[115, 124, 2746, 5382, 8623]

In [121]:
# Create the train and test dataframe

train = pd.DataFrame()
test = pd.DataFrame()

# Set up the train size
train_size = 0.8

for user_id in user_ids1000:
    if train1000.loc[[user_id]].shape[0] <=2:
        print(user_id)
        continue
    elif train1000.loc[[user_id]].shape[0] > 2: 
        df = train1000.loc[[user_id]]
        n = df.shape[0]
        test_start_index = round(train_size * n)
        df_train = df.iloc[:test_start_index]
        df_test = df.iloc[test_start_index:]     
        train = pd.concat([train, df_train])
        test = pd.concat([test, df_test])
    
# Print the shape of the original, train and test
train1000.shape, train.shape, test.shape

1880240
2148001


((296615, 9), (237322, 9), (59290, 9))

In [124]:
# Take a peek at the train dataset
train.head(2)

row_id  timestamp  content_id  content_type_id  task_container_id  \
user_id                                                                      
115           0          0        5692                0                  1   
115           1      56943        5716                0                  2   

         user_answer  answered_correctly  prior_question_elapsed_time  \
user_id                                                                 
115                3                   1                          0.0   
115                2                   1                      37000.0   

        prior_question_had_explanation  
user_id                                 
115                              False  
115                              False

In [125]:
train.tail(2)

row_id  timestamp  content_id  content_type_id  task_container_id  \
user_id                                                                      
5577726  296611  625327721        4303                0                  8   
5577726  296612  625344560        4445                0                  9   

         user_answer  answered_correctly  prior_question_elapsed_time  \
user_id                                                                 
5577726            3                   0                      10000.0   
5577726            0                   1                       6000.0   

        prior_question_had_explanation  
user_id                                 
5577726                           True  
5577726                           True

In [126]:
# Take a peek at the test
test.head(2)

row_id  timestamp  content_id  content_type_id  task_container_id  \
user_id                                                                      
115          37    1415188        7882                0                 37   
115          38    1468285        7962                0                 38   

         user_answer  answered_correctly  prior_question_elapsed_time  \
user_id                                                                 
115                2                   1                      20000.0   
115                3                   1                      17000.0   

        prior_question_had_explanation  
user_id                                 
115                               True  
115                               True

In [127]:
test.tail(2)

row_id  timestamp  content_id  content_type_id  task_container_id  \
user_id                                                                      
5577726  296613  625367161        5206                0                 10   
5577726  296614  625378759        3928                0                 11   

         user_answer  answered_correctly  prior_question_elapsed_time  \
user_id                                                                 
5577726            1                   0                       6000.0   
5577726            0                   0                      13000.0   

        prior_question_had_explanation  
user_id                                 
5577726                           True  
5577726                           True

### Count how many lectures each user has reviewed in the train

In [128]:
# Create a new columns for if the student review the lecture or not

train['reviewed_lecture'] = (train['answered_correctly'] == -1)
train

row_id  timestamp  content_id  content_type_id  task_container_id  \
user_id                                                                      
115           0          0        5692                0                  1   
115           1      56943        5716                0                  2   
115           2     118363         128                0                  0   
115           3     131167        7860                0                  3   
115           4     137965        7922                0                  4   
...         ...        ...         ...              ...                ...   
5577726  296608      95083        5153                0                  5   
5577726  296609     118668        5266                0                  6   
5577726  296610  625311764        5328                0                  7   
5577726  296611  625327721        4303                0                  8   
5577726  296612  625344560        4445                0                  9   

         user_answer  answered_correctly  prior_question_elapsed_time  \
user_id                                                                 
115                3                   1                          0.0   
115                2                   1                      37000.0   
115                0                   1                      55000.0   
115                0                   1                      19000.0   
115                1                   1                      11000.0   
...              ...                 ...                          ...   
5577726            1                   1                      14000.0   
5577726            3                   0                      20000.0   
5577726            0                   0                      20000.0   
5577726            3                   0                      10000.0   
5577726            0                   1                       6000.0   

        prior_question_had_explanation  reviewed_lecture  
user_id                                                   
115                              False             False  
115                              False             False  
115                              False             False  
115                              False             False  
115                              False             False  
...                                ...               ...  
5577726                          False             False  
5577726                          False             False  
5577726                          False             False  
5577726                           True             False  
5577726                           True             False  

[237322 rows x 10 columns]

In [130]:
# Compute how many times each user reviewed the lecture

count_reviewed_lecture = train.groupby(train.index).reviewed_lecture.sum()

user_id
115        0
124        0
2746       1
5382       1
8623       2
          ..
5555086    0
5566848    2
5568049    1
5571864    0
5577726    0
Name: reviewed_lecture, Length: 998, dtype: int64

**Takeaways**
- What if the user didn't review the lecture in the train dataset but reference the lecture in the test dataset? 
- For this qucik run though, I didn't use it as the new feature

### Drop the lecture rows in train and test

In [131]:
# Drop the lecture rows in train

mask = train['answered_correctly'] != -1
train = train[mask]

# Print the shape
train.shape

(232718, 10)

**Quick note**: the number of observations in the train drops from 237322 to 232718

In [132]:
# Drop the lecture rows in test

mask = test['answered_correctly'] != -1
test = test[mask]

# Print the shape
test.shape

(58222, 9)

**Quick note**: the number of observations in the test drops from 59290 to 58222.

### Feature engineering - how to measure the whole content
- content_id
- task_container_id

In [135]:
# Compute the statistics of the content by the the content_id

content_stats = train.groupby('content_id').answered_correctly.agg(['mean', 'count', 'std', 'median', 'skew'])


# Rename the name of the columns

content_stats.columns = ['mean_accuracy', 'question_asked', 'std_accuracy', 'median_accuracy', 'skew_accuracy']

# Take a look at the dataframe
content_stats

mean_accuracy  question_asked  std_accuracy  median_accuracy  \
content_id                                                                 
0                0.944444              18      0.235702              1.0   
1                0.928571              14      0.267261              1.0   
2                0.566372             113      0.497783              1.0   
3                0.793103              58      0.408619              1.0   
4                0.480519              77      0.502897              0.0   
...                   ...             ...           ...              ...   
13518            0.000000               1           NaN              0.0   
13519            0.000000               1           NaN              0.0   
13520            0.750000               4      0.500000              1.0   
13521            0.000000               1           NaN              0.0   
13522            0.000000               2      0.000000              0.0   

            skew_accuracy  
content_id                 
0               -4.242641  
1               -3.741657  
2               -0.271474  
3               -1.485842  
4                0.079539  
...                   ...  
13518                 NaN  
13519                 NaN  
13520           -2.000000  
13521                 NaN  
13522                 NaN  

[12373 rows x 5 columns]

In [136]:
# Compute the statistics of the content by the the task_container_id

task_container_stats = train.groupby('task_container_id').answered_correctly.agg(['mean', 
                                                                                  'count', 
                                                                                  'std', 
                                                                                  'median', 
                                                                                  'skew'])


# Rename the name of the columns

task_container_stats.columns = ['mean_accuracy', 
                                'question_asked', 
                                'std_accuracy', 
                                'median_accuracy', 
                                'skew_accuracy']

# Take a look at the dataframe
task_container_stats

mean_accuracy  question_asked  std_accuracy  \
task_container_id                                                
0                       0.696215            1004      0.460120   
1                       0.522908            1004      0.499724   
2                       0.429429             999      0.495243   
3                       0.543912            1002      0.498317   
4                       0.493096            1738      0.500096   
...                          ...             ...           ...   
4496                    1.000000               1           NaN   
4497                    1.000000               1           NaN   
4498                    1.000000               1           NaN   
4499                    1.000000               1           NaN   
4500                    0.000000               1           NaN   

                   median_accuracy  skew_accuracy  
task_container_id                                  
0                              1.0      -0.854589  
1                              1.0      -0.091867  
2                              0.0       0.285566  
3                              1.0      -0.176595  
4                              0.0       0.027644  
...                            ...            ...  
4496                           1.0            NaN  
4497                           1.0            NaN  
4498                           1.0            NaN  
4499                           1.0            NaN  
4500                           0.0            NaN  

[4501 rows x 5 columns]